### DISCUSSION TITLE
export values for types

### DISCUSSION THREAD

**Discussion by vogler** [prisma/prisma#5291] [Q&A] [answered]
I wanted to have simpler `update`/`delete` which just take `data` and handle `where`.
The problem is that `update`/`delete` fail if excess fields are passed as `where`.
I could explicitly list the unique fields like below, but I wanted to build something generic.
Example (frontend code which imports `@prisma/client` of server):

~~~typescript
const _update = (args => rest('PUT', args)) as Prisma.TodoDelegate['update'];
export const update = (data: Prisma.TodoWhereUniqueInput) => _update({ data, where: { id: data.id } })
~~~

To be able to 'cast' `data` to `TodoWhereUniqueInput` and also limit its fields to it, I'd need some value for the type.
The following is already generated:
~~~typescript
 export const TodoScalarFieldEnum: {
    id: 'id',
    createdAt: 'createdAt',
    updatedAt: 'updatedAt',
    text: 'text',
    done: 'done'
  };
...
 export type TodoWhereUniqueInput = {
    id?: number
  }
~~~
So, for the case of `..WhereUniqueInput`, generate something like
~~~typescript
  export const TodoWhereUniqueInputFields = ['id'];
~~~

To have more information available at runtime would be helpful for all kinds of generic definitions (error handling, automatic REST API etc.).

**Comment by vogler** [comment]
There's no value for `PrismaAction`:
~~~typescript
type Prisma.PrismaAction = "findUnique" | "findFirst" | "findMany" | "create" | "createMany" | "update" | "updateMany" | "upsert" | "delete" | "deleteMany" | "executeRaw" | "queryRaw" | "aggregate" | "count"
~~~
So currently, I have to list the actions in my code:
~~~typescript
const actions = ['findMany', 'create', 'update', 'delete', 'findUnique', 'findFirst', 'updateMany', 'upsert', 'deleteMany', 'aggregate', 'count'] as const; // these are the actions defined on each model. TODO get from prisma? PrismaAction is just a type.
~~~

**Comment by vogler** [comment, marked as answer]
Instead of just generating the types, you could use some existing library for a run-time representation.
https://github.com/colinhacks/zod seems popular.
It by default strips keys which are not in the schema: https://github.com/colinhacks/zod#unrecognized-keys
Having extra keys on objects is currently problematic since Prisma errors and there is no run-time 'cast', so you have to resort to weird boilerplate duplicating the information already present in the type.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
